src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1916 @@
+                // certain things (i.e. escape references) and what errors will be reported.
+                //
+                // This is necessary as .net regexes allow references to *future* captures.
+                // As such, we don't know when we're seeing a reference if it's to something
+                // that exists or not.
+                var tree1 = new RegexParser(text, options,
+                    ImmutableDictionary<string, TextSpan>.Empty,
+                    ImmutableDictionary<int, TextSpan>.Empty).ParseTree();
+
+                var analyzer = new CaptureInfoAnalyzer(text);
+                var (captureNames, captureNumbers) = analyzer.Analyze(tree1.Root, options);
+
+                var tree2 = new RegexParser(
+                    text, options, captureNames, captureNumbers).ParseTree();
+                return tree2;
+            }
+            catch (Exception e) when (StackGuard.IsInsufficientExecutionStackException(e))
+            {
+                return null;
+            }
+        }
+
+        private RegexTree ParseTree()
+        {
+            var expression = this.ParseAlternatingSequences(consumeCloseParen: true);
+            Debug.Assert(_lexer.Position == _lexer.Text.Length);
+            Debug.Assert(_currentToken.Kind == RegexKind.EndOfFile);
+
+            var root = new RegexCompilationUnit(expression, _currentToken);
+
+            var diagnostics = ArrayBuilder<EmbeddedDiagnostic>.GetInstance();
+            CollectDiagnostics(root, diagnostics);
+
+            return new RegexTree(
+                _lexer.Text, root, diagnostics.ToImmutableAndFree(),
+                _captureNamesToSpan, _captureNumbersToSpan);
+        }
+
+        private static void CollectDiagnostics(RegexNode node, ArrayBuilder<EmbeddedDiagnostic> diagnostics)
+        {
+            foreach (var child in node)
+            {
+                if (child.IsNode)
+                {
+                    CollectDiagnostics(child.Node, diagnostics);
+                }
+                else
+                {
+                    var token = child.Token;
+                    foreach (var trivia in token.LeadingTrivia)
+                    {
+                        AddUniqueDiagnostics(trivia.Diagnostics, diagnostics);
+                    }
+
+                    AddUniqueDiagnostics(token.Diagnostics, diagnostics);
+                }
+            }
+        }
+
+        /// <summary>
+        /// It's very common to have duplicated diagnostics.  For example, consider "((". This will
+        /// have two 'missing )' diagnostics, both at the end.  Reporting both isn't helpful, so we
+        /// filter duplicates out here.
+        /// </summary>
+        private static void AddUniqueDiagnostics(ImmutableArray<EmbeddedDiagnostic> from, ArrayBuilder<EmbeddedDiagnostic> to)
+        {
+            foreach (var diagnostic in from)
+            {
+                if (!to.Contains(diagnostic))
```

Comment:
I'm worried about time complexity here. Would it make sense to use a hash set for de-dupe?

Reply:
sure.  i can do that.

Reply:
#resolved